extensions/avro-extensions/src/main/java/io/druid/data/input/AvroStreamInputRowParser.java

```diff
@@ -0,0 +1,116 @@
+import io.druid.data.input.impl.TimestampSpec;
+import io.druid.data.input.schemarepo.SubjectAndIdConverter;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.util.ByteBufferInputStream;
+import org.joda.time.DateTime;
+import org.schemarepo.Repository;
+import org.schemarepo.api.TypedSchemaRepository;
+import org.schemarepo.api.converter.AvroSchemaConverter;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.List;
+
+public class AvroStreamInputRowParser<SUBJECT, ID> implements ByteBufferInputRowParser
+{
+  private final ParseSpec parseSpec;
+  private final List<String> dimensions;
+
+  private final TypedSchemaRepository<ID, Schema, SUBJECT> typedRepository;
+  private final SubjectAndIdConverter<SUBJECT, ID> subjectAndIdConverter;
+  private final Repository schemaRepository;
+
+  @JsonCreator
+  public AvroStreamInputRowParser(
+      @JsonProperty("parseSpec") ParseSpec parseSpec,
+      @JsonProperty("subjectAndIdConverter") SubjectAndIdConverter<SUBJECT, ID> subjectAndIdConverter,
+      @JsonProperty("schemaRepository") Repository schemaRepository
+  )
+  {
+    this.parseSpec = parseSpec;
+    this.dimensions = parseSpec.getDimensionsSpec().getDimensions();
+    this.subjectAndIdConverter = subjectAndIdConverter;
+    this.schemaRepository = schemaRepository;
+    typedRepository = new TypedSchemaRepository<ID, Schema, SUBJECT>(
+        schemaRepository,
+        subjectAndIdConverter.getIdConverter(),
+        new AvroSchemaConverter(false),
+        subjectAndIdConverter.getSubjectConverter()
+    );
+  }
+
+  @Override
+  public InputRow parse(ByteBuffer input)
+  {
+    ByteBufferInputStream inputStream = new ByteBufferInputStream(Collections.singletonList(input));
+    Pair<SUBJECT, ID> subjectAndId = subjectAndIdConverter.toSubjectAndId(input);
+    Schema schema = typedRepository.getSchema(subjectAndId.lhs, subjectAndId.rhs);
+    DatumReader<GenericRecord> reader = new GenericDatumReader<GenericRecord>(schema);
+    try {
+      GenericRecord record = reader.read(null, DecoderFactory.get().binaryDecoder(inputStream, null));
+      return doParse(record, parseSpec, dimensions);
+    }
+    catch (IOException e) {
+      throw new RuntimeException("Fail to decode avro message!", e);
+    }
+  }
+
+  protected static InputRow doParse(
+      GenericRecord record, ParseSpec parseSpec, List<String> dimensions
+  )
+  {
+    TimestampSpec timestampSpec = parseSpec.getTimestampSpec();
+    DateTime dateTime = timestampSpec.extractTimestamp(
+        Collections.singletonMap(
+            timestampSpec.getTimestampColumn(),
+            record.get(timestampSpec.getTimestampColumn())
+        )
+    );
+    return new MapBasedInputRow(dateTime, dimensions, new GenericRecordAsMap(record));
```

Comment:
@himanshug I think ARRAY is already handled because ARRAY is expressed as `org.apache.avro.generic.GenericData#Array` which extends `AbstractList`. For UNION, can we just check the actual type? For instance field type is ["null","int","array"], if the actual field value is int, then return `int`, if array, then return `Array`? I'm not using pig AvroStorage, maybe I didn't understand the problem, correct me if I'm wrong please.